File: softpos-evolution-java-light/app/src/main/java/it/nexi/softpos_evo_java_light/activities/PaymentInsertValuesActivity.java
package it.nexi.softpos_evo_java_light.activities;

import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.google.gson.Gson;

import java.util.HashMap;
import java.util.Map;

import it.nexi.softpos_evo_java_light.databinding.ActivityPaymentInsertValuesBinding;
import it.nexi.softpos_evo_java_light.domain.Constant;
import it.nexi.softpos_evo_java_light.domain.LoadingDialog;
import it.nexi.softpos_evo_java_light.services.SDKNexiService;
import it.nexi.softpos_evo_java_light.services.Utils;
import it.nexipos.app2app.data.model.PaymentData;


/**
 *
 * Nexi Payment
 *
 * Manages the insertion of values to be used for the execution
 * of the Nexi POS Payment method
 *
 */
public class PaymentInsertValuesActivity extends AppCompatActivity {

    private ActivityPaymentInsertValuesBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);

        binding = ActivityPaymentInsertValuesBinding.inflate(getLayoutInflater());

        setContentView(binding.getRoot());

        Gson gson = new Gson();

        // Set random value
        String callerTrxIdRandom = Utils.getRandomCallerTrxId();
        binding.editTextCallerTrxId.setText(callerTrxIdRandom);

        String amountRandom = Utils.getRandomAmount();
        binding.editTextAmount.setText(amountRandom);

        binding.btnStartPayment.setOnClickListener(v -> {

            boolean autoClose   = binding.checkBoxAutoClose.isChecked();
            boolean isUrlTicket = binding.checkBoxUrlTicket.isChecked();
            boolean hideRetry   = binding.checkBoxHideRetry.isChecked();
            boolean isTipAmount = binding.checkBoxTipEnabled.isChecked();

            String amount       = binding.editTextAmount.getText().toString();
            String email        = binding.editTextEmail.getText().toString();
            String sms          = binding.editTextSms.getText().toString();
            String callerTrxId  = binding.editTextCallerTrxId.getText().toString();
            String tipAmount    = binding.editTextTipAmount.getText().toString();

            String addInfo1     = binding.editTextAddInfo1.getText().toString();
            String addInfo2     = binding.editTextAddInfo2.getText().toString();
            String addInfo3     = binding.editTextAddInfo3.getText().toString();
            String addInfo4     = binding.editTextAddInfo4.getText().toString();
            String addInfo5     = binding.editTextAddInfo5.getText().toString();

            if (amount.isEmpty()) {

                Utils.showInfoAlert(this,"Error","'amount' cannot be empty");

                return;

            }

            if (callerTrxId.isEmpty()) {

                Utils.showInfoAlert(this,"Errore","'callerTrxId' cannot be empty");

                return;

            }

            boolean isSendTicket = !autoClose;

            final String amountForPos     = Utils.convertAmountForPos(amount);
            final String tipAmountForPos  = tipAmount.isEmpty() ? "" : Utils.convertAmountForPos(tipAmount);

            // Creating the data object for the sdk
            PaymentData paymentData = getPaymentData(amountForPos,
                                                     callerTrxId,
                                                     isSendTicket,
                                                     isUrlTicket,
                                                     Constant.CALLER_NAME,
                                                     addInfo1,
                                                     addInfo2,
                                                     addInfo3,
                                                     addInfo4,
                                                     addInfo5,
                                                     email,
                                                     sms,
                                                     autoClose,
                                                     hideRetry,
                                                     isTipAmount,
                                                     tipAmountForPos
                                                    );

            String jsonPayment = gson.toJson(paymentData);

            Log.d(Constant.TAG_SOFTPOS_EVO_JAVA, "PaymentData ->  " + jsonPayment);

            // starts the sdk method.
            startPayment(paymentData);

        }); // end binding.btnStartPayment.setOnClickListener(v -> {

        // An Observer is instantiated to receive communication from the Nexi sdk,
        // after the sdk's method is launched.
        SDKNexiService.getOperationResult().observe(this, result -> {

            // SDKNexiService communicated the outcome of the
            // operation to the activity
            if (result == null) return;

            LoadingDialog.hide();

            if (result.isSuccess()) {

                // SDKNexiService returned a successful result.

                Toast.makeText(this, "OK: " + result.getDescription(), Toast.LENGTH_LONG).show();

                // The activity is closed and the SDK
                // launches the Nexi POS, passing it the
                // parameters to execute the requested operation.
                finish();

            } else {

                // SDKNexiService returned an error.
                Utils.showInfoAlert(this, "Error", result.getDescription());

            }

            // Resets the LiveData state to prevent the same message
            // from being forwarded multiple times
            // (for example, in the event of a smartphone orientation change).
            SDKNexiService.resetResult();

        }); // end SDKNexiService.getOperationResult().observe(this, result -> {

    } // end onCreate(

    /**
     *
     * Creating the data object for the sdk
     *
     */
    private PaymentData getPaymentData(String amount,
                                       String callerTrxId,
                                       boolean isSendTicket,
                                       boolean isUrlTicket,
                                       String callerName,
                                       String addInfo1,
                                       String addInfo2,
                                       String addInfo3,
                                       String addInfo4,
                                       String addInfo5,
                                       String email,
                                       String sms,
                                       boolean autoClose,
                                       boolean hideRetry,
                                       boolean isTipAmount,
                                       String tipAmount
                                      ) {

        Map<String, String> mapAddInfo = new HashMap<>();

        if (addInfo1 != null && !addInfo1.isEmpty()) mapAddInfo.put("addInfo1", addInfo1);
        if (addInfo2 != null && !addInfo2.isEmpty()) mapAddInfo.put("addInfo2", addInfo2);
        if (addInfo3 != null && !addInfo3.isEmpty()) mapAddInfo.put("addInfo3", addInfo3);
        if (addInfo4 != null && !addInfo4.isEmpty()) mapAddInfo.put("addInfo4", addInfo4);
        if (addInfo5 != null && !addInfo5.isEmpty()) mapAddInfo.put("addInfo5", addInfo5);

        return new PaymentData(amount,
                               callerTrxId,
                               isSendTicket,
                               isUrlTicket,
                               isTipAmount,
                               tipAmount,
                               callerName,
                               mapAddInfo,
                               email,
                               sms,
                               autoClose,
                               hideRetry
                              );

    } // end getPaymentData(

    /**
     *
     * Starts the sdk method
     *
     */
    private void startPayment(PaymentData data) {

        LoadingDialog.show(this);

        // requests the SDKNexiService to start the method
        SDKNexiService.executePay(data);

    } // end private void startPayment(PaymentData data) {

} // end Activity

File: softpos-cloud-auth-demo-android-kotlin/app/src/main/java/it/nexi/softpos_evo_kotlin_01/activities/PaymentInsertValuesActivity.kt
package it.nexi.softpos_evo_kotlin_01.activities

import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import it.nexi.softpos_evo_kotlin_01.domain.Constant
import it.nexi.softpos_evo_kotlin_01.commons.LoadingDialog
import it.nexi.softpos_evo_kotlin_01.services.SDKNexiService
import it.nexi.softpos_evo_kotlin_01.commons.showInfoAlert
import it.nexi.softpos_evo_kotlin_01.databinding.ActivityPaymentInsertValuesBinding
import it.nexipos.app2app.data.model.PaymentData
import kotlinx.coroutines.launch

/**
 *
 * Nexi Payment
 *
 * Manages the insertion of values to be used for the execution
 * of the Nexi POS Payment method
 *
 */
class PaymentInsertValuesActivity : AppCompatActivity() {

    private lateinit var binding: ActivityPaymentInsertValuesBinding

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)

        binding = ActivityPaymentInsertValuesBinding.inflate(layoutInflater)

        setContentView(binding.root)

        with(binding) {

            btnStartPayment.setOnClickListener {

                val autoClose   = checkBoxAutoClose.isChecked
                val isUrlTicket = checkBoxUrlTicket.isChecked
                val hideRetry   = checkBoxHideRetry.isChecked
                val rawAmount   = editTextAmount.text.toString()
                val email       = editTextEmail.text.toString()
                val sms         = editTextSms.text.toString()
                val callerTrxId = editTextCallerTrxId.text.toString()

                val addInfo1    = editTextAddInfo1.text.toString()
                val addInfo2    = editTextAddInfo2.text.toString()
                val addInfo3    = editTextAddInfo3.text.toString()
                val addInfo4    = editTextAddInfo4.text.toString()
                val addInfo5    = editTextAddInfo5.text.toString()

                val isSendTicket = !autoClose

                var finalAmount = ""

                val amountDouble = rawAmount.toDoubleOrNull() ?: 0.0

                if (amountDouble > 0.0) {

                    val formatted = String.format(java.util.Locale.ITALY, "%.2f", amountDouble)

                    finalAmount = formatted.replace(".", "").replace(",", "")
                }

                if (finalAmount.isEmpty() || callerTrxId.isEmpty()) {

                    showInfoAlert(
                        context = this@PaymentInsertValuesActivity,
                        titolo = "Mandatory fields",
                        messaggio = "amount and callerTrxId, cannot be empty."
                    )

                    return@setOnClickListener

                }

                // Creating the data object for the sdk
                val paymentData  = getPaymentData(finalAmount,
                                                  callerTrxId,
                                                  isSendTicket,
                                                  isUrlTicket,
                                                Constant.CALLER_NAME,
                                                  addInfo1,
                                                  addInfo2,
                                                  addInfo3,
                                                  addInfo4,
                                                  addInfo5,
                                                  email,
                                                  sms,
                                                  autoClose,
                                                hideRetry
                                                 )

                // starts the sdk method.
                startPayment(paymentData)

            } // btnStartPayment.setOnClickListener {

        } // with

        // awaits the response from the SDKNexiService
        lifecycleScope.launch {

            SDKNexiService.operationResult.collect { result ->

                LoadingDialog.hide()

                result?.let {

                    if (it.isSuccess) {

                        Toast.makeText(this@PaymentInsertValuesActivity,
                                    "OK: ${it.description}",
                                        Toast.LENGTH_LONG).show()

                        // The activity is closed and the SDK
                        // launches the Nexi POS, passing it the
                        // parameters to execute the requested operation.
                        finish()

                    } else {

                        showInfoAlert(this@PaymentInsertValuesActivity,
                                "Error",
                                        it.description
                                     )

                    }

                    // Resets the LiveData state to prevent the same message
                    // from being forwarded multiple times
                    // (for example, in the event of a smartphone orientation change).
                    SDKNexiService.resetResult()

                } // result?.let {

            } //  PaymentManager.operationResult.collect { result ->

        } //  lifecycleScope.launch {

    } // onCreate(

    /**
     *
     * Creating the data object for the sdk
     *
     */
    private fun getPaymentData(amount: String,
                               callerTrxId: String,
                               isSendTicket: Boolean,
                               isUrlTicket: Boolean,
                               callerName: String,
                               addInfo1: String,
                               addInfo2: String,
                               addInfo3: String,
                               addInfo4: String,
                               addInfo5: String,
                               email: String,
                               sms: String,
                               autoClose: Boolean,
                               hideRetray : Boolean
                              ) : PaymentData {

        val mapAddInfo = mutableMapOf<String, String>()

        if (!addInfo1.isNullOrEmpty()) mapAddInfo["addInfo1"] = addInfo1
        if (!addInfo2.isNullOrEmpty()) mapAddInfo["addInfo2"] = addInfo2
        if (!addInfo3.isNullOrEmpty()) mapAddInfo["addInfo3"] = addInfo3
        if (!addInfo4.isNullOrEmpty()) mapAddInfo["addInfo4"] = addInfo4
        if (!addInfo5.isNullOrEmpty()) mapAddInfo["addInfo5"] = addInfo5

        return PaymentData(amount,
                           callerTrxId,
                           isSendTicket,
                           isUrlTicket,
                           callerName,
                        mapAddInfo,
                           email,
                           sms,
                           autoClose,
                           hideRetray
                          )

    } // fun getPaymentData

    /**
     *
     * Starts the sdk method
     *
     */
    private fun startPayment(data: PaymentData) {

        lifecycleScope.launch {

            LoadingDialog.show(this@PaymentInsertValuesActivity)

            // requests the SDKNexiService to start the method
            SDKNexiService.executePay(data)

        } // lifecycleScope.launch {

    } // fun startPayment

} // end Activity